### PR TITLE
Cypress e2e testing

### LINKED ISSUE #227: Convert demos to actual unit tests
* Release number or git hash: v4.0.0-alpha.12
 * Web browser version and OS: any

### Observed behavior
No proper unit/UI testing in place. 

### Expected behavior
Demos are useful for manually checking if everything works correctly, but unit tests ensure correctness. 
**It would also make collaboration easier as side effects can be detected more easily.**

### PR DESCRIPTION
**Pull request by sgratzl**
closes #227 depends on #279 

**prerequisites**: 
 * [x] branch is up-to-date with the branch to be merged with, i.e. develop
 * [x] build is successful
 * [x] code is cleaned up and formatted 
 * [ ] tested with Firefox 52, Firefox 57+, Chrome 64+, MS Edge 16+

### Summary
 
wip of using cypress to run e2e tests. 

**Notes**

* as part of this the location of the typings have changed to `/build/src`
* the idea is to test close to the UMD case. So, as if one uses the builder. Thus, the bundled/compiled version of LineUp.js is used.

#### Variant 1: with prebuilt LineUp

```sh
npm run compile
npm run build:prod
npm run cy:open
```

#### Variant 2: with webpack-dev-server

First shell:

```sh
npm run compile
npm start
```

Second shell:

```sh
npm run cy:start
```

### CODE REVIEW

**Comment by sgratzl** [comment]
local testing

```sh
npm run compile
npm run build:prod
npm run cy:open
```

**Comment by sgratzl** [comment]
I added a simple test for #275 we can add more tests on the fly

File: cypress/integration/big.spec.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import {withLineUp, waitReady} from './utils/lineup';
+import {generateData, DEFAULT_CATEGORIES} from './utils/data';
+
+it('builder2', withLineUp((lineUpJS, document) => {
```
Comment (thinkh) [inline cypress/integration/big.spec.ts:4]: This test is missing the wrapped `describe()` function.

File: .circleci/config.yml [hunk 2]
```diff
@@ -7,14 +7,16 @@ jobs:
...
+          keys:
+            - v2-deps-{{ checksum "package.json" }}-{{ checksum "package-lock.json" }}
+            - v2-deps-{{ checksum "package.json" }}
+            - v2-deps-
+      - run: npm ci
       - save_cache:
-          key: deps3-{{ checksum "package.json" }}-{{ checksum "package-lock.json" }}
+          key: v2-deps-{{ checksum "package.json" }}-{{ checksum "package-lock.json" }}
           paths:
-          - ./node_modules
```
Comment (thinkh) [inline .circleci/config.yml:17]: Why is it not necessary anymore to cache the node_modules directory?

File: cypress/integration/builder2.spec.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import {setupLineUp, waitReady} from './utils/lineup';
+import {generateData, DEFAULT_CATEGORIES} from './utils/data';
+
+it('builder2', async () => {
```
Comment (thinkh) [inline cypress/integration/builder2.spec.ts:4]: This test is missing the wrapped `describe()` function.

File: .circleci/config.yml (hunk 2 above)
Comment (sgratzl) [reply .circleci/config.yml:17]: take a look at the other cache paths that directly caches the npm cache

**Comment by sgratzl** [comment]
> @sgratzl How should we proceed with this PR? Should I wait for further implementation or merge it once you addressed my two minor change requests?

merge then we can start using it

**Pull request merged**